File: 3509.maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.py
#
# @lc app=leetcode id=3509 lang=python3
#
# [3509] Maximum Product of Subsequences With an Alternating Sum Equal to K
#

# @lc code=start
from typing import List
class Solution:
    def maxProduct(self, nums: List[int], k: int, limit: int) -> int:
        from functools import lru_cache
        n = len(nums)
        res = -1

        @lru_cache(maxsize=None)
        def dp(i, alt_sum, parity, prod, used):
            nonlocal res
            if i == n:
                if used > 0 and alt_sum == k and prod <= limit:
                    res = max(res, prod)
                return
            # Skip current element
            dp(i+1, alt_sum, parity, prod, used)
            # Include current element, update parity and alternating sum
            new_prod = prod * nums[i]
            if new_prod <= limit:
                if parity == 0:
                    next_alt_sum = alt_sum + nums[i]
                else:
                    next_alt_sum = alt_sum - nums[i]
                dp(i+1, next_alt_sum, 1-parity, new_prod, used+1)
        dp(0, 0, 0, 1, 0)
        return res
# @lc code=end

File: 3509.maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.java
#
# @lc app=leetcode id=3509 lang=java
#
# [3509] Maximum Product of Subsequences With an Alternating Sum Equal to K
#
# @lc code=start
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxProduct(int[] nums, int k, int limit) {
        int n = nums.length;
        // Use a HashMap for memoization to avoid out-of-bounds with alternating sum
        Map<String, Integer> memo = new HashMap<>();
        int[] res = new int[]{-1};
        dfs(nums, k, limit, 0, 0, 0, 1, false, memo, res);
        return res[0];
    }

    void dfs(int[] nums, int k, int limit, int i, int sum, int parity, int prod, boolean taken, Map<String, Integer> memo, int[] res) {
        if (prod > limit) return;
        if (i == nums.length) {
            if (taken && sum == k) {
                if (prod > res[0]) res[0] = prod;
            }
            return;
        }
        // Key as string to handle all possible sum values
        String key = i + "," + sum + "," + parity;
        if (memo.containsKey(key) && memo.get(key) >= prod) return;
        memo.put(key, prod);
        // Skip current element
        dfs(nums, k, limit, i+1, sum, parity, prod, taken, memo, res);
        // Include current element
        int newSum = parity == 0 ? sum + nums[i] : sum - nums[i];
        dfs(nums, k, limit, i+1, newSum, 1-parity, prod * nums[i], true, memo, res);
    }
}
# @lc code=end